ingester/ingester.go

```diff
@@ -536,55 +566,61 @@ func (i *Ingester) flushLoop(j int) {
 		}
 		op := o.(*flushOp)
 
-		// get the user
-		i.userStateLock.Lock()
-		userState, ok := i.userState[op.userID]
-		i.userStateLock.Unlock()
-		if !ok {
-			continue
+		if err := i.flushUserSeries(op.userID, op.fp, op.immediate); err != nil {
+			log.Errorf("Failed to flush user: %v", err)
 		}
-		ctx := user.WithID(context.Background(), op.userID)
+	}
+}
 
-		// Decide what chunks to flush
-		series, ok := userState.fpToSeries.get(op.fp)
-		if !ok {
-			continue
-		}
+func (i *Ingester) flushUserSeries(userID string, fp model.Fingerprint, immediate bool) error {
+	i.userStateLock.Lock()
+	userState, ok := i.userState[userID]
+	i.userStateLock.Unlock()
+	if !ok {
+		return nil
+	}
 
-		userState.fpLocker.Lock(op.fp)
+	series, ok := userState.fpToSeries.get(fp)
+	if !ok {
+		return nil
+	}
 
-		// Assume we're going to flush everything
-		chunks := series.chunkDescs
+	userState.fpLocker.Lock(fp)
+	if !i.shouldFlushSeries(series, immediate) {
+		userState.fpLocker.Unlock(fp)
+		return nil
+	}
 
-		// If the head chunk is old enough, close it
-		if op.immediate || model.Now().Sub(series.lastTime) > i.cfg.MaxChunkIdle {
-			series.closeHead()
-		} else {
-			chunks = chunks[:len(chunks)-1]
-		}
-		userState.fpLocker.Unlock(op.fp)
+	// Assume we're going to flush everything, and maybe don't flush the head chunk if it doesn't need it.
+	chunks := series.chunkDescs
+	if immediate || (len(chunks) > 0 && i.shouldFlushChunk(chunks[0])) {
+		series.closeHead()
+	} else {
+		chunks = chunks[:len(chunks)-1]
+	}
+	userState.fpLocker.Unlock(fp)
```

Comment:
I've lost track of how Go does references / mutation, but IIUC, we only need to keep this lock to resolve the condition on the `if` statement, not to process the body. If this is right, do you think it would be clearer to do something like:

```golang
chunks := series.chunkDescs
shouldFlushHead := len(chunks) > 0 && i.shouldFlushChunk(chunks[0])
userState.fpLocker.Unlock(fp)
if immediate || shouldFlushHead {
  // ...
```

?